### DISCUSSION TITLE
在资源管理器中展开显示具体文件中的内容

### DISCUSSION THREAD

**Discussion by lingtaol** [opensumi/core#2500] [General]
有一个需求，想在资源管理器中的某个具体的文件，展开文件具体的内容。比如json文件，我想在资源管理器中展开显示package.json文件中的依赖项。类似于这样：
![image](https://user-images.githubusercontent.com/29857969/227833281-6b6f6ad8-23dc-4d02-9e25-4ff25df47dd8.png)

有什么好的意见吗，目前有两个方案：
1. VsCode TreeView API， 想了一下，还是要在资源管理中显示，不能重新开个视图显示。
2. 修改File Tree Next的源码，不清楚怎么改

**Comment by erha19** [comment]
@ringtu 修改 FileTreeNext 源码方案可以看这里的代码：
https://github.com/opensumi/core/blob/ad721ea4259a2305f2287d82e792ae9331abc0dd/packages/file-tree-next/src/browser/services/file-tree-api.service.ts#L33-L67

这里的逻辑是文件树在渲染时最终请求到的数据位置，你可以在下面代码位置追加一个判断 https://github.com/opensumi/core/blob/ad721ea4259a2305f2287d82e792ae9331abc0dd/packages/file-tree-next/src/browser/services/file-tree-api.service.ts#L101

```ts
if (filestat.isDirectory && filestat.url.endwith('package.json')) {
  // ...
}
```

同时在 https://github.com/opensumi/core/blob/ad721ea4259a2305f2287d82e792ae9331abc0dd/packages/file-tree-next/src/browser/services/file-tree-api.service.ts#L41 额外 处理这种情况下的返回值即可。

**Reply by erha19** [reply]
另外，你这里也不需要直接修改源码去发包，维护会有些困难，建议直接通过我们的 DI 机制去 override , 如框架默认的注册是：

```ts
{
  token: IFileTreeAPI,
  useClass: FileTreeAPI,
},
```

你这里把 `FileTreeAPI` 重新实现后，通过在 RenderAPP 阶段执行类似下面的语句即可：

```ts
injector.overrideProviders({
  token: IFileTreeAPI,
  useClass: {你自己实现的 Class}
})
```

@ringtu

**Reply by lingtaol** [reply]
@erha19 好的，谢谢，我尝试一下。

还有一个额外的问题：
问题描述：我现在用的是ide-electron项目来开发， 项目中有需要用到jar包，我将其放至了根目录下。我如何在Node模块中获取到该jar包的路径，有没有什么库可以用的

**Reply by erha19** [reply]
你可以把你的 jar 包放进 resources 目录下面，打包脚本内会在打包阶段将这个目录带进产物，需要你根据打包前后的路径自行处理一下不同环境下的资源路径拼接，不需要引入额外的库，你使用相对路径引入就行了

**Comment by lingtaol** [comment]
我根据上面要求修改了resolveChildren()， 可以生成对应的树，但是生成的树结构有点问题。
![image](https://user-images.githubusercontent.com/29857969/228186123-49d0e806-9119-4185-bf6f-cbed8bfe0fa5.png)
我在创建Directory中已经指定了name， 但是真实目录或文件名可能会根据uri去刷新？

**Reply by erha19** [reply]
@ringtu 文件树上为了实现通过 URI 反查节点 path 的能力，采用了 uri 作为每个文件节点的唯一值去进行的节点渲染，而 Tree 组件渲染时需要保证 path 的唯一性，以 File 节点为例：

https://github.com/opensumi/core/blob/e81a1cfc5b86627e389d8d95de6dcf692a1a635a/packages/file-tree-next/src/common/file-tree-node.define.ts#L53-L62

这里传入的 `{name}` 就是用于节点路径拼接的路径，你这里需要保证每个传入的 name 的唯一性，这是文件树实现上面所做的限制。

**Reply by erha19** [reply]
举个例子，假设这里有一个  `a/b/c.js` 文件，那 A 节点传入的 `name=a`,  最终的 path 为  `root/a`, B 节点传入的 `name=b`,  最终的 path 为  `root/a/b` 以此类推。

**Reply by lingtaol** [reply]
但是我的文件树不是真实的文件，都是XCPsim.a2l 这个文件中的内容树，这种情况下，我如何去设置我们uri和name

**Reply by erha19** [reply]
@ringtu 你在构建你的自定义节点的时候不传就行了，例如你只是将原有的 File 节点 `extends CompositeTreeNode`，另外新增一个 `FileContentNode extends TreeNode`, 这里面的构造函数不要传入 name，直接 `super(tree as ITree, parent)`。

同时，文件树内还有一些地方对于 uri 有强依赖，你也需要处理一下，看到报错的时候处理好就行了。

**Reply by lingtaol** [reply]
我现在自定义了节点，发现如果没有传入uri和filestat，渲染时会报错，报错信息如下：
![image](https://user-images.githubusercontent.com/29857969/228419969-96a393e6-547a-4c98-a922-52c34a0abb31.png)
不知道这个怎么解决

**Reply by erha19** [reply]
你需要自己定位到具体的依赖逻辑，if(!xx.url) return 之类的处理一下，这个已经不是框架能处理的了，是你自己业务逻辑需要去自行处理的。
